### DISCUSSION TITLE
The heuristic to subtract 1 million gas to estimate max counts is OK

### DISCUSSION THREAD

**Discussion by PaulRBerg** [sablier-labs/evm-monorepo#932] [Show and tell]
## Context

I said [here](https://github.com/sablier-labs/v2-core/pull/908#issuecomment-2115959358) that I will follow up on the cross-chain gas usage issue.

## My Input

It turns out that the heuristic introduced by @smol-ninja in https://github.com/sablier-labs/v2-core/pull/908, i.e. to subtract 1 million gas from the block gas limit, accounts for the whole gamut of cross-chain differences in Minimum Transaction Gas. So we can keep it like that.

I did some comprehensive research today, and it appears that:

1. No chain has a Minimum Transaction Go greater than 1 million
2. Execution costs are mostly equivalent to Ethereum Mainnet.

The only exception to point no. 2 is zkSync, which provides gas refunds. However, this is not a problem because it means that we will _overestimate_, not _underestimate_. In other words, our estimated max counts will be more than enough to cover the actual execution cost on zkSync.

## Gas Fee Models

| Chain     | Type              | Min Tx Gas | Model                     | Explanation                                                          |
| --------- | ----------------- | ---------- | ------------------------- | -------------------------------------------------------------------- |
| Arbitrum  | Optimistic Rollup | 35-70k     | Multifunctional gas limit | Bake L1 data fee into L2 gas limit                                   |
| BNB Chain | Sidechain         | 21k        | EVM                       | Same as vanilla EVM                                                  |
| Gnosis    | Sidechain         | 21k        | EVM                       | Same as vanilla EVM                                                  |
| OP Stack  | Optimistic Rollup | 21k        | Two-dimensional fees      | Explicit L2 execution fee and implicit L1 DA fee                     |
| Polygon   | Sidechain         | 21k        | EVM                       | Same as vanilla EVM                                                  |
| Scroll    | ZK Rollup         | 21k        | Two-dimensional fees      | Explicit L2 execution fee and implicit L1 DA fee                     |
| zkSync    | ZK Rollup         | 120k-190k  | Custom: AA + state diffs  | Gas costs fluctuate a lot due to Account Abstraction and state diffs |

## References

- https://github.com/mds1/evm-diff/issues/31
- https://ethereum.stackexchange.com/q/163222/24693

---

cc @sablier-labs/solidity

[^1]: OP Stack includes Optimism and Base

**Comment by smol-ninja** [comment]
Thank you for the research on this subject. This is a very good summary tbh. 👏👏👏

**Reply by PaulRBerg** [reply]
Cheers.
